src/OpenLoco/src/Map/MapGenerator.cpp

```diff
@@ -392,6 +400,112 @@ namespace OpenLoco::World::MapGenerator
         }
     };
 
+    class HeightMapTerrainGenerator
+    {
+    public:
+        void generate(HeightMapRange heightMap)
+        {
+            generateFromHeightmapPng(heightMap);
+        }
+
+    private:
+        void getPixel(const std::vector<png_byte>& imageData, int x, int y, png_byte& red, png_byte& green, png_byte& blue, png_byte& alpha, int width)
+        {
+            int index = (y * width + x) * 4;
+            red = imageData[index];
+            green = imageData[index + 1];
+            blue = imageData[index + 2];
+            alpha = imageData[index + 3];
+        }
+
+        bool openUiPngBrowser()
+        {
+            auto basePath = Platform::getUserDirectory();
+
+            // open ui prompt window
+            // doesn't work...
+            bool confirm = Ui::Windows::PromptBrowse::open(OpenLoco::Ui::Windows::PromptBrowse::browse_type::load, (char*)basePath.string().c_str(), "*.png", StringIds::title_prompt_load_game);
+
+            return confirm;
+        }
+
+        // note: png image must be 384x384 or otherwise this method will probably crash
+        int generateFromHeightmapPng(HeightMapRange heightMap)
+        {
+            auto basePath = Platform::getUserDirectory();
+            auto defaultHeightmap = basePath / "heightmap.png";
+            auto filename = defaultHeightmap.string();
+
+            // todo:
+            /*if (openUiPngBrowser())
+            {
+                filename = std::string(_savePath);
+            }*/
+
+            std::ifstream inFile(filename, std::ios::binary);
+
+            png_structp png = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+            if (!png)
+            {
+                std::cerr << "Failed to create PNG read struct" << std::endl;
+                return 1;
+            }
+
+            png_infop info = png_create_info_struct(png);
+            if (!info)
+            {
+                std::cerr << "Failed to create PNG info struct" << std::endl;
+                png_destroy_read_struct(&png, NULL, NULL);
+                return 1;
+            }
+
+            // Set error handler
+            if (setjmp(png_jmpbuf(png)))
+            {
+                png_destroy_read_struct(&png, &info, NULL);
+                throw std::runtime_error("PNG ERROR");
+            }
+
+            FILE* file = fopen(filename.c_str(), "rb");
```

Comment:
This is not utf8 aware. You will need to use ifstream and set the png read function afaik. Or use `_wfopen` on windows but i'd prefer we didn't have platform specific code.

Reply:
fixed